app/src/main/java/org/oppia/android/app/mydownloads/downloads/DownloadsFragmentPresenter.kt

```diff
@@ -0,0 +1,130 @@
+    binding.apply {
+      this.lifecycleOwner = fragment
+      this.viewModel = downloadsViewModel
+    }
+    downloadsViewModel.setInternalProfileId(internalProfileId)
+    downloadsViewModel.setProfileId(internalProfileId)
+    binding.downloadsRecyclerView.apply {
+      adapter = createRecyclerViewAdapter()
+    }
+    return binding.root
+  }
+
+  private fun createRecyclerViewAdapter(): BindableAdapter<DownloadsItemViewModel> {
+    return BindableAdapter.MultiTypeBuilder
+      .newBuilder<DownloadsItemViewModel, ViewType> { downloadsItemViewModel ->
+        when (downloadsItemViewModel) {
+          is DownloadsSortByViewModel -> ViewType.VIEW_TYPE_SORT_BY
+          is DownloadsTopicViewModel -> ViewType.VIEW_TYPE_TOPIC
+          else -> throw IllegalArgumentException(
+            "Encountered unexpected view model: $downloadsItemViewModel"
+          )
+        }
+      }
+      .registerViewDataBinder(
+        viewType = ViewType.VIEW_TYPE_SORT_BY,
+        inflateDataBinding = DownloadsSortbyBinding::inflate,
+        setViewModel = this::bindDownloadsSortBy,
+        transformViewModel = { it as DownloadsSortByViewModel }
+      )
+      .registerViewDataBinder(
+        viewType = ViewType.VIEW_TYPE_TOPIC,
+        inflateDataBinding = DownloadsTopicCardBinding::inflate,
+        setViewModel = this::bindDownloadsTopicCard,
+        transformViewModel = { it as DownloadsTopicViewModel }
+      )
+      .build()
+  }
+
+  private enum class ViewType {
+    VIEW_TYPE_SORT_BY,
+    VIEW_TYPE_TOPIC
+  }
+
+  private fun bindDownloadsSortBy(
+    binding: DownloadsSortbyBinding,
+    viewModel: DownloadsSortByViewModel
+  ) {
+    binding.viewModel = viewModel
+    val items = listOf("Newest", "Alphabetical", "Download Size", "Option 4")
+    val adapter = ArrayAdapter(fragment.requireContext(), R.layout.downloads_sortby_menu, items)
+    binding.sortByMenu.setInputType(0)
+    binding.sortByMenu.setText(adapter.getItem(0).toString(), false)
+    binding.sortByMenu.setAdapter(adapter)
+  }
+
+  private fun bindDownloadsTopicCard(
+    binding: DownloadsTopicCardBinding,
+    viewModel: DownloadsTopicViewModel
+  ) {
+    binding.viewModel = viewModel
+
+    var isDeleteExpanded = false
+    binding.isDeleteExpanded = isDeleteExpanded
+
+    binding.expandListIcon.setOnClickListener {
+      isDeleteExpanded = when (isDeleteExpanded) {
+        true -> {
+          false
+        }
+        else -> {
+          true
+        }
+      }
+      binding.isDeleteExpanded = isDeleteExpanded
+    }
+
+    binding.deleteImageView.setOnClickListener {
+      downloadsViewModel.isAllowedDownloadAccess.observe(
+        fragment,
+        Observer { isAllowedDownloadAccess ->
+          val dialogFragment = DownloadsTopicDeleteDialogFragment
+            .newInstance(internalProfileId, isAllowedDownloadAccess)
+          dialogFragment.showNow(fragment.childFragmentManager, TAG_DOWNLOAD_TOPIC_DELETE_DIALOG)
```

Comment:
@rt4914 bit concerned about opening dialog while observing the data. Could you suggest a good approach here?

Reply:
I think this is obsolete now?